Better handling of (de)serialization of circular and forward references with natural keys

matthijskooijman opened the pull request:
This PR addresses [#26552](https://code.djangoproject.com/ticket/26552) and [#31051](https://code.djangoproject.com/ticket/31051), both of which address problems with (de)serializing data involving forward or circular references and natural keys.

charettes commented on django/core/management/commands/dumpdata.py:
```diff
@@ -144,7 +144,16 @@ def get_objects(count_only=False):
             """
-            models = serializers.sort_dependencies(app_list.items())
+            if use_natural_foreign_keys:
+                models = serializers.sort_dependencies(app_list.items(), allow_loops=True)
```
If I understand correctly _cycles_ might be a better term than _loops_ here.

matthijskooijman replied:
Agreed, I'll change that.

matthijskooijman replied:
Done, will push soon.

charettes commented on django/core/management/commands/dumpdata.py:
```diff
@@ -144,7 +144,16 @@ def get_objects(count_only=False):
+                    if app_models == None:
+                        models.extend(app_config.get_models())
+                    else:
+                        models.extend(app_models)
```
Any reason why `serializers.sort_dependencies(app_list.items(), allows_cycles=False)` wouldn't work here?

If that's the case then this whole chunk could be replaced by

```python
serializers.sort_dependencies(
    app_list.items(), allow_cycles=use_natural_foreign_keys
)

matthijskooijman replied:
That would also work, but do sorting work that is not useful at all. I guess it just comes down to efficiency then.

charettes commented on tests/test_runner_apps/circular_ref/models.py:
```diff
@@ -0,0 +1,39 @@
+
+# Module containing a self-reference, and with natural keys
+class SelfRefThing(models.Model):
+    key = models.CharField(max_length=100)
```
`key` should likely be unique if it's used as a natural key?

matthijskooijman replied:
Good point. I based this off an existing model that also does not specify uniqueness: https://github.com/django/django/blob/c33eb6dcd0c211f8f02b2976fe3b3463f0a54498/tests/fixtures/models.py#L121-L124

Enforcing uniqueness is not relevant to the testcase (that is, things are unique in the testdata, which is sufficient), so keeping it like this might be more consistent. Alternatively, I can make the existing model also unique, that might actually be a good idea.

matthijskooijman replied:
Done, I made changed the existing model touched by the commit as well. There's still a lot of other natural key models without uniqueness in other tests, but I think that is out of scope, so I'm resolving this. I will push the changes soon.

felixxm replied:
I cleaned it up in #12653.

charettes commented on django/test/runner.py:
```diff
@@ -617,6 +618,7 @@ def get_test_runner_kwargs(self):
             'resultclass': self.get_resultclass(),
             'verbosity': self.verbosity,
+            'stream': self.stream,
         }
```
It's not necessary to add this publicly available feature, that would need to be documented and tested. Simply subclass `DiscoverRunner` and override `get_test_runner_kwargs`.

matthijskooijman replied:
Ah, good plan. Willdo.

matthijskooijman replied:
Done, will push soon.

charettes commented on tests/test_runner/tests.py:
```diff
@@ -468,3 +469,25 @@ def test_run_checks_passes_and_teardown_raises(self):
+    # these models, to check that deserializing such instances with forward references is not problematic anymore.
+    def test_setup_databases(self):
+        with modify_settings(INSTALLED_APPS={'append': 'test_runner_apps.circular_ref'}):
+            from test_runner_apps.circular_ref.models import CircularRefThingA, CircularRefThingB, SelfRefThing
```
`test_runner_apps` should likely live in `test_runner.app_something` instead. Also this import should be moved to the top.

matthijskooijman replied:
`test_runner_apps` already exists, so moving that seems out of scope for this PR.

IIRC the import cannot be at the top, since then a model is imported that is not in `INSTALLED_APPS`, which I think caused a problem. I'll doublecheck and add a comment.

matthijskooijman replied:
Turns out the import was actually unused, so I removed it and will resolve this conversation. If you want to continue discussing moving `test_runner_apps`, please add a new comment to the PR for that.

charettes commented on django/db/backends/base/creation.py:
```diff
@@ -130,8 +131,15 @@ def deserialize_db_from_string(self, data):
+        # Run without constraint checks (which requires a transaction), so any forward references or even loops in the
+        # data can be loaded without problems.
+        with atomic(using=self.connection.alias):
+            with self.connection.constraint_checks_disabled():
```
SQLite cannot disable constraints within a transaction

https://github.com/django/django/blob/c33eb6dcd0c211f8f02b2976fe3b3463f0a54498/django/db/backends/sqlite3/base.py#L289-L296

So you'll likely need another solution here.

matthijskooijman replied:
Looks like it works at the start of a transaction, right?

charettes replied:
It has to be issued before a transaction is opened. In short `'PRAGMA foreign_keys = OFF'` is a silent noop after a `BEGIN` has been issued.

charettes replied:
More details about it from https://sqlite.org/pragma.html

charettes replied:
It happens to work here because all base backends except for MySQL use deferred foreign keys (MySQL doesn't support them) so no checks are performed until the transaction commits.

matthijskooijman replied:
Oh, interesting, then `loaddata` also has it in the wrong order. I wonder if MySQL, which also uses this method, can have the order swapped or not... Also, I think I have not seen test failures on Sqlite even with this order, so maybe sqlite already defers constraint checks to the end of a transaction and offers `disable_constraint_checking` to disable constraint checking outside of a transaction? This needs some additional investigation and testing, I guess.

matthijskooijman replied:
So, for this particular case, it is ok that `disable_constraint_checking()` fails inside a transaction, since the transaction itself already provides the needed semantics. I guess this does mean that the API of `disable_constraint_checking()` and `constraint_checks_disabled()` is a bit ill-defined, but maybe that is out of scope for this PR.

So I would propose keeping the code as-is, since it should work for the purpose? @charettes, would you agree?

charettes replied:
@matthijskooijman lets keep it as is if `loaddata` uses this pattern. I guess we could eventually add a connection level database backend specified manager that hides this implementation detail but this should do in the mean time.

matthijskooijman commented:
Thanks for your quick review. I've responded inline to everything, most things I will need to change. I'm still hammering out the last few kinks to keep the full testsuite working, once I manage that and fixed your remarks, I'll update the PR.

matthijskooijman commented:
In the meanwhile, I was writing some more considerations wrt this PR. I was planning on posting these after running the full testsuite, but since I ran into some problems and people started reviewing, I'll post this comment already.


With regard to the actual fixes, I am pretty confident that these are ok. The only issue is about backward compatibility for `django.serializers.core.sort_dependencies()`. This function used to throw an exception when a circular dependency involving natural keys was present (e.g. when it could not produce a perfectly valid ordering), but now also allows doing a best-effort sorting while still allowing cyclic dependencies using a `allow_loops` argument. This argument defaults to `False`, for compatibility with existing callers. However, with this PR, only one caller is present within django, that sets `allow_loop` to True. I guess the bottomline is: Is this functoin considered public API and should it keep compatibility? The function is not mentioned in the docs AFAICS.

With regard to the testcases, I think they work but there is some room for improvement.

In particular, with the added models to the fixtures tests, when the problem that is tested for occurs, it is likely that this problem occurs in other testcases too (since part of the problem prevents serializing of any app containing a model, not even an instance, with circular references with natural keys). This is just a minor problem, when the bug does come back, it will certainly be detected by some test(s). It could be fixed by moving these models into a separate app that is only used by this particular test, but that seems a hassle for little gain.

With regard to the new test_runner test, it re-initialized the default database, which could slightly influence other tests (in practice, it should leave a flushed database with some extra data in the serialized copy stored on the database connection, which should be ok?).

I tried using a mock for `django.utils.connections` (as seen elsewhere) to make the testrunner use a new, separate, database, but that seemed to properly call `create_test_db` for the mocked database, but then the subsequent migration data would not show up in the serialized contents, leading me to believe that the `migrate` done by `create_test_db` happened on the original default database (since that uses `db.connections`). I then tried mocking `django.db.connections` instead, leading to a `no such table: django_migrations`. Then mocking both lead to `database table is locked`, so I'm not quite sure what the right approach is here...

Here's what I tried (in an older version of the test, but I think the same should hold for the test as it is now):

{{{
class FooTest(unittest.TestCase):
    def test_setup_databases(self):
        tested_connections = db.ConnectionHandler({
            'default': {
                'ENGINE': 'django.db.backends.sqlite3',
                'NAME': ':memory:',
            },
        })
        with mock.patch('django.db.connections', new=tested_connections):
            with mock.patch('django.test.utils.connections', new=tested_connections):
                runner_instance = DiscoverRunner(verbosity=0)
                old_config = runner_instance.setup_databases()
                runner_instance.teardown_databases(old_config)
        print(tested_connections['default']._test_serialized_contents)
}}}

Any pointers on how to approach this, or is there perhaps an existing test that does something similar that I overlooked?

matthijskooijman commented:
> With regard to the new test_runner test, it re-initialized the default database, which could slightly influence other tests, which might be harmless enough. However, on the postgres, postgis and mysql backends, this tries to recreate the same database, giving e.g. psycopg2.errors.DuplicateDatabase: database "test_django_38_pull-requests-bionic" already exists (see djangoci output).
>
> I tried using a mock for django.utils.connections (as seen elsewhere) to make the testrunner use a new, separate, database

Thinking a bit more about this, I guess it will not be generally possible to actually create a new database. For sqlite, creating a new file or in-memory database could work, but for external database servers, there is no guarantee that you actually have permission to create another database (and if so, what name that should have).

When running multiple parallel testrunners, cloning already makes additional databases, but those have predictable names and if the test environment does not support additional databases, you can disable that with `--parallel=1`.

An alternative fix for this would be to pass `keepdb` to the test runner, so it would not error out on an existing database, but instead just reuse it. I've just force pushed this (along with fixes to most of @charettes comments) and the first mysql tests seem to have passed, so that's promising, but postgres is still failing with "psycopg2.errors.UndefinedTable: relation "postgres_tests_testjsonmodel" does not exist
LINE 1: ...gres_tests_testjsonmodel"."jsonmodel_ptr_id" FROM "postgres_..." for some reason...

matthijskooijman commented:
I had another look at this and could reproduce the JSON-related error locally. What I suspect happens is that some database content from earlier tests remains in the database, which then causes a problem with the serialization in `CircularRefTestRunnerTest`, or something like that.

What I also noticed, is that both Postgresql and Mysql locally try to access the `test_test_django_main` database, so the `test_` prefix is added twice (probably once by the main test runner and once by the test runner inside `CircularRefTestRunnerTest`).

In other words, reusing the same database is tricky as well. This leads to another solution to this problem: Add a third database to the test settings. IOW, next to the "default" and "other" database, add e.g. a "testrunnertest" database (name up for suggestions). This database would *not* be declared in the `databases=...` of `CircularRefTestRunnerTest`, so it will not be automatically created on startup of the testrunner, leaving it free to be created from scratch *and* teared down by `CircularRefTestRunnerTest` (and maybe other tests that want to test database creation and/or teardown). Of course, this requires changes to the test settings modules used by people, so this test must only run if such a database is actually available.

I have not tried setting this up yet, but how does this sound?

haricot commented:
''



'

Le mer. 18 déc. 2019 à 21:35, Matthijs Kooijman <notifications@github.com>
a écrit :

> I had another look at this and could reproduce the JSON-related error
> locally. What I suspect happens is that some database content from earlier
> tests remains in the database, which then causes a problem with the
> serialization in CircularRefTestRunnerTest, or something like that.
>
> What I also noticed, is that both Postgresql and Mysql locally try to
> access the test_test_django_main database, so the test_ prefix is added
> twice (probably once by the main test runner and once by the test runner
> inside CircularRefTestRunnerTest).
>
> In other words, reusing the same database is tricky as well. This leads to
> another solution to this problem: Add a third database to the test
> settings. IOW, next to the "default" and "other" database, add e.g. a
> "testrunnertest" database (name up for suggestions). This database would
> *not* be declared in the databases=... of CircularRefTestRunnerTest, so
> it will not be automatically created on startup of the testrunner, leaving
> it free to be created from scratch *and* teared down by
> CircularRefTestRunnerTest (and maybe other tests that want to test
> database creation and/or teardown). Of course, this requires changes to the
> test settings modules used by people, so this test must only run if such a
> database is actually available.
>
> I have not tried setting this up yet, but how does this sound?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/django/django/pull/12166?email_source=notifications&email_token=AACUDLMU3Y627NPB7RGK6G3QZKCSRA5CNFSM4JTOTSNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHHNEVI#issuecomment-567202389>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACUDLLVMLTMD5JTHJME33TQZKCSRANCNFSM4JTOTSNA>
> .
>

haricot commented:
LM
nkwkwkwkxk‹‹'f''g'f'f'f'jjbbbbb
Bnjjjlkklkkkkk

Le mer. 18 déc. 2019 à 21:35, Matthijs Kooijman <notifications@github.com>
a écrit :

> I had another look at this and could reproduce the JSON-related error
> locally. What I suspect happens is that some database content from earlier
> tests remains in the database, which then causes a problem with the
> serialization in CircularRefTestRunnerTest, or something like that.
>
> What I also noticed, is that both Postgresql and Mysql locally try to
> access the test_test_django_main database, so the test_ prefix is added
> twice (probably once by the main test runner and once by the test runner
> inside CircularRefTestRunnerTest).
>
> In other words, reusing the same database is tricky as well. This leads to
> another solution to this problem: Add a third database to the test
> settings. IOW, next to the "default" and "other" database, add e.g. a
> "testrunnertest" database (name up for suggestions). This database would
> *not* be declared in the databases=... of CircularRefTestRunnerTest, so
> it will not be automatically created on startup of the testrunner, leaving
> it free to be created from scratch *and* teared down by
> CircularRefTestRunnerTest (and maybe other tests that want to test
> database creation and/or teardown). Of course, this requires changes to the
> test settings modules used by people, so this test must only run if such a
> database is actually available.
>
> I have not tried setting this up yet, but how does this sound?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/django/django/pull/12166?email_source=notifications&email_token=AACUDLMU3Y627NPB7RGK6G3QZKCSRA5CNFSM4JTOTSNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHHNEVI#issuecomment-567202389>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACUDLLVMLTMD5JTHJME33TQZKCSRANCNFSM4JTOTSNA>
> .
>

matthijskooijman commented:
> What I also noticed, is that both Postgresql and Mysql locally try to access the test_test_django_main database, so the test_ prefix is added twice

Seems this was a separate issue: https://code.djangoproject.com/ticket/31117 (or at least that issue has the same symptom). No longer relevant here, though.

> In other words, reusing the same database is tricky as well. This leads to another solution to this problem: Add a third database to the test settings.

I tried this as a solution for https://code.djangoproject.com/ticket/31117 (see https://github.com/django/django/pull/12247), but that turned out to be quite complicated and also fragile.

So, I settled for a less complete testcase: Instead of testing the full test-runner flow, include the database setup, serializing db contents, running testcases, deserializing and teardown, the testcase now just picks out the problematic parts (serializing and deserializing) and tests those. This makes the testcases a bit less general than before, but sufficient to verify that these particular bugs are fixed.

With that, I am happy with these fixes and testcases, so they are ready for review and merging IMHO.

felixxm commented on tests/backends/base/test_creation.py:
```diff
@@ -67,3 +70,45 @@ def test_migrate_test_setting_true(self, mocked_migrate, mocked_ensure_connectio
+
+    def test_serialize_circular_ref(self):
+        import django
+        print(django.conf.settings.INSTALLED_APPS)
```
Please remove debug prints.

matthijskooijman replied:
W00ps, seems I missed one. Removed it :-)

matthijskooijman commented:
@felixxm good suggestion, I created #12453 for that. It includes the fix commit from this PR, and I added a new testcase, since the testcase I had only worked when both issues were fixed. I've also rebased this PR on top of master and to include that new PR, but let's focus on the other PR first.

charettes commented:
Not related to the latest changes but I wonder how much time is actually spent by `deserialize_db_from_string` to perform the JSON deserialization given `data = serializers.serialize('python', get_objects())` could be used to store the object representation while avoiding the continued round trip.

matthijskooijman commented:
@charettes I'm not sure what you mean there? Are you suggesting to change the implementation of `serialize_db_to_string` to use python serialization instead of JSON as a performance improvement? Or is this about this PR directly?

charettes commented:
@matthijskooijman it isn't directly about this PR.

> Are you suggesting to change the implementation of serialize_db_to_string to use python serialization instead of JSON as a performance improvement?

I was wondering if that would make a difference. It's surely faster to operate from a list of Python object than converting a string to it before hand. The question is how significant that is given all the `Model.save` involved anyway.

matthijskooijman commented:
> I was wondering if that would make a difference.

Hm, I'm not so sure. I guess it also depends on the size of the data - AFAIU this is only about data generated during migration and post-migrated, so usually just contenttypes and permissions, I think.

matthijskooijman commented:
@felixxm Thanks for merging. I rebased the fix for ticket-31051 on top of your improvements that are now in master, fixing some trivial conflicts and simplifying the testcase a bit (it also needs no new `connection` object AFAICS). I pushed that to a separate PR for clarity (#12458), making this PR obsolete.
